ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs

```diff
@@ -0,0 +1,306 @@
+  , Result (..)
+  , completeApplicationTx
+
+  -- * Traces
+  , ErrorPolicyTrace (..)
+  , traceErrorPolicy
+  , WithAddr (..)
+
+  -- * Re-exports of PeerState
+  , PeerStates
+  , SuspendDecision (..)
+  ) where
+
+import           Control.Exception (Exception, SomeException (..))
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Foldable (traverse_)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe, mapMaybe)
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Tracer
+
+import           Data.Semigroup.Action
+
+import           Ouroboros.Network.Subscription.PeerState
+
+data ErrorPolicy where
+     ErrorPolicy :: forall err.
+                      Exception err
+                   => (ConnectionOrApplicationException err -> SuspendDecision DiffTime)
+                   -> ErrorPolicy
+
+instance Show ErrorPolicy where
+    show (ErrorPolicy (_err :: ConnectionOrApplicationException err -> SuspendDecision DiffTime)) =
+           "ErrorPolicy ("
+        ++ tyConName (typeRepTyCon (typeRep (Proxy :: Proxy err)))
+        ++ ")"
+
+evalErrorPolicy :: forall e.
+                   Exception e
+                => ConnectionOrApplicationException e
+                -> ErrorPolicy
+                -> Maybe (SuspendDecision DiffTime)
+evalErrorPolicy e p =
+    case p of
+      ErrorPolicy (f :: ConnectionOrApplicationException e' -> SuspendDecision DiffTime)
+        -> case gcast e :: Maybe (ConnectionOrApplicationException e') of
+              Nothing -> Nothing
+              Just e' -> Just $ f e'
+
+-- | Evaluate a list of 'ErrorPolicy's; If none of them applies this function
+-- returns 'Nothing', in this case the exception will be traced and not thrown.
+--
+evalErrorPolicies :: forall e.
+                     Exception e
+                  => ConnectionOrApplicationException e
+                  -> [ErrorPolicy]
+                  -> Maybe (SuspendDecision DiffTime)
+evalErrorPolicies e =
+    f . mapMaybe (evalErrorPolicy e)
+  where
+    f :: [SuspendDecision DiffTime]
+      -> Maybe (SuspendDecision DiffTime)
+    f []          = Nothing
+    f (cmd : rst) = Just $ sconcat (cmd :| rst)
+
+
+-- | List of error policies for exception handling and a policy for handing
+-- application return values.
+--
+data ErrorPolicies m addr a = ErrorPolicies {
+    epErrorPolicies  :: [ErrorPolicy]
+  , epReturnCallback :: Time m -> addr -> a -> SuspendDecision DiffTime
```

Comment:
@coot Can you explain what this is? What should consensus do?

Reply:
This let one do something when a protocol returns, we decided that in `consensus` we'd rather throw an exception so I don't think you'll need to do here anything: I think I need to modify the return type to `Maybe (SuspendDecision DiffTime)` (or remove the callback all together).

Btw, it's only possible for client side to finish any of the protocols we have.